Rewrite of 'Add Folder To Project'.

dennykorsukewitz opened the pull request:
<!--
The manual review may take several days or weeks,
depending on the reviewer's availability and workload.
Patience padawan!

You can request a review from @packagecontrol-bot.
Please ensure the reviews pass and follow any instructions.

Please provide some information via this checklist,
feel free to remove what't not applicable.
-->

- [x] I'm the package's author and/or maintainer.
- [x] I have have read [the docs][1].
- [ ] I have tagged a release with a [semver][2] version number.
- [x] My package repo has a description and a README describing what it's for and how to use it.
- [ ] My package doesn't add context menu entries. *
- [x] My package doesn't add key bindings. **
- [x] Any commands are available via the command palette.
- [x] Preferences and keybindings (if any) are listed in the menu and the command palette, and open in split view.
- [ ] If my package is a syntax it doesn't also add a color scheme. ***
- [ ] If my package is a syntax it is named after the language it supports (without suffixes like "syntax" or "highlighting").
- [x] I use [.gitattributes][3] to exclude files from the package: images, test files, sublime-project/workspace.

---
Hi guys, 
my package is [Sublime-AddFolderToProject](https://github.com/dennykorsukewitz/Sublime-AddFolderToProject)


**AddFolderToProject** is a Sublime Text package that streamlines the process of adding and removing folders to and from your projects. It provides a set of commands and context menu options to manage your project's folder structure directly from the editor.


This extension already started as a small function in the  [Znuny-Sublime package](https://github.com/znuny/Znuny-Sublime/blob/dev/ZnunyAddToProject.py).
However, I had to extend this function and place it in an extra package.

**My package is similar to AddFolderToProject-SublimePlugin:**
I noticed that there is already an [AddFolderToProject-SublimePlugin](https://github.com/DavidGerva/AddFolderToProject-SublimePlugin) package. 

**However it should still be added because:**
But apparently this is no longer maintained since 09.02.2015. Issues and PullRequest were not further processed.

I have compiled and revised my functions and the functions of my colleague. 
I have solved all known problems and now want to make it available.

Of course, I don't want to displace my colleague's work. That's why I've given the package the 2.
Or should we replace the existing one? 


I will tag a release if nothing speaks against this merge from your side.

### ChannelRepositoryTools: Test Default Channel
```
Running Default Channel Tests
  ----------------------------------------------------------------------
  Ran 9682 tests in 0.175s
  
  OK
```

---

<!-- 
*)   Unless it definitely really needs them,
     they apply to the cursor's context
     and their visibility is conditional.
     Space in this menu is limited!
**)  There aren't enough keys for all packages,
     you'd risk overriding those of other packages.
     You can put commented out suggestions in a keymap file, 
     and/or explain how to create bindings in your README.
***) We have hundreds of color schemes,
     and plenty of scopes to make any syntax work. 

For bonus points also considered how the review guidelines apply to your package:
https://github.com/wbond/package_control_channel/wiki#reviewing-a-package-addition

For updates to existing packages:
If your package isn't using tag based releases,
please switch to tags now.
 -->

[1]: https://packagecontrol.io/docs/submitting_a_package
[2]: https://semver.org
[3]: https://www.git-scm.com/docs/gitattributes#_export_ignore

deathaxe commented:
If an unmaintained old package already exists with nearly the same (or less) functionality, a valid option was to replace it (or take over its name).

Another but way simpler package to note is [AddRemoveFolder](https://github.com/fsaad/Sublime-AddRemoveFolder).

With all the functionality of your package however a name like "Folder Manager" was probably a better fit.

Actually I like the idea of having a single or limited amount of high quality packages, which handle all major aspects of managing folders and cooperate well with packages like [Project Manager](https://packagecontrol.io/packages/ProjectManager) or [Project and Workspace Management](https://packagecontrol.io/packages/Project%20And%20Workspace%20Management).

Other recently updated packages with focus to handling folders are
- a [Folder Aliaser](https://packagecontrol.io/packages/Folder%20Aliaser) to change folders' labels.
- a [AutoProjects](https://github.com/SublimeText/AutoProjects) to auto-generate sublime-project files for folders opened via `subl <folder>`.

> [!NOTE]
>
> I was already created a command to remove folders, too, which makes use of ST's InputHandlers for a more seamless and more modern looking integration into Command Palette.
>
> ![grafik](https://github.com/wbond/package_control_channel/assets/16542113/97028ded-bf12-4255-9eb5-e0242fc50518)

---

With out-of-the-box existing `Project: Add Folder` command, I wonder if it would also make sense to group commands of your package under `Project:` for usability-consistency reasons, e.g.:

- Project: Add current file's folder
- Project: Add current file's parent folder  
  _note: the default "Project: Add Folder" opens a dialog box to choose any folder_
- Project: Remove Folder
- ...

---

A command `Copy Path` is already available via context menu in side bar. It copies a selected folder's or file's absolute path. So the commands in your package are redundant.

Packages like `File Manager` also provide more variants of them.

---

Any new plugin must opt-in to python 3.8 plugin_host via [.python-version](https://github.com/SublimeText/FolderAliaser/blob/master/.python-version).

dennykorsukewitz commented:
Hey  @deathaxe ,
thanks for all the feedback.
I've seen some of the other packages, but you've also shown me a few new ones.

If we replace the existing one, I would have to do the following:

- [ ] [.python-version](https://github.com/SublimeText/FolderAliaser/blob/master/.python-version).
- [ ] change code to use of ST's InputHandlers 
- [ ] group commands under Project: for usability-consistency reasons
- [ ] remove obsolete `Copy Path` command
- [ ] release

But now I'm honestly confused as to what I should do best. What would be your suggestion?

Best regards

braver commented:
Sorry for letting you dangle for a while there. Been busy with.. "life" I guess 😅 And I must say this thread had be a little bit confused as well. 

> Or should we replace the existing one?

I would say yes. "Add Folder To Project" is positively ancient, and we can see there are already various other approaches to covering parts (or more) of that feature set, a new replacement for that existing package sounds like a good move to me. We of course need to check with the authors of that package to see if they object, and if maybe you can simply contribute your work to that existing repo. And then make sure you message the user base about what happened. 

> I wonder if it would also make sense to group commands of your package under Project: 

I can see how that would be nice, also because `AddFolderToProject` is a bit of a long name. But you could argue that the labelling is also there to indicate which package is providing the feature. So perhaps leave it as is (and consider a shorter name for your package 😉).

> remove obsolete Copy Path command
> [.python-version](https://github.com/SublimeText/FolderAliaser/blob/master/.python-version).

Those sound like good (and easy) improvements.

> change code to use of ST's InputHandlers

You could do that at a later point.

TLDR; first step would be to consider replacing, or contributing to, the original package by @DavidGerva

braver commented:
Please respond to the feedback to continue the review.

dennykorsukewitz commented:
Hi, I'll give you more feedback tomorrow.

dennykorsukewitz commented:
Hi @braver 

I wrote to David @DavidGerva again.

```
Hi David,

first of all thanks for your OpenSource work!
Especially in the SublimePlugin area.

https://github.com/DavidGerva/AddFolderToProject-SublimePlugin

I've been using your AddFolderToProject plugin for quite a long time, but I still had requests for improvements.
Unfortunately, you are no longer active in this repo and therefore current points were no longer processed. 

See PR https://github.com/DavidGerva/AddFolderToProject-SublimePlugin/pull/1

I have now incorporated my own extensions into AddFolderToProject https://github.com/wbond/package_control_channel/pull/8904 and revised and updated your existing functions. However, understandably there should not be a second AddFolderToProject plugin.

Now I'm standing around a bit. The plan would be to replace your plugin with mine.
The same functions as before, just a few more.

But we can't just do this without your permission. The great respect among us developers should be preserved.

Would you give us permission to do this? Or how do you feel about it?

Kind regards,
Denny

```

I hope we get an answer.
As soon as I have received feedback, I will let you know.

P.S. I understand about living life ;)

braver commented:
@dennykorsukewitz thanks! Hope that works out.

DavidGerva commented:
For what it is worth, LGTM! :)

dennykorsukewitz commented:
> @dennykorsukewitz thanks! Hope that works out.

He has already replied:

```
Hi Denny!
Sure, go ahead.

And thank you for your work :)
David
```

braver commented:
Alright, then replacing the existing package is the way forward. Please make the required changes here, and tag release with a higher number than the last release of the old package. It’s always a good idea to link back to the old repo from your repo, and explain the takeover. If you intend to also make changes that are visible to the end user, like changing the command palette entry labels, you might want to consider creating a message for your new version to explain that to any existing users.

dennykorsukewitz commented:
## Todo:

- [x] Please make the required changes here. https://github.com/wbond/package_control_channel/pull/8904/commits/0559a45e45ff22bde4a4ef78797c280aae4ae46a
- [ ] tag release with a higher number than the last release of the old package.
- [ ] It’s always a good idea to link back to the old repo from your repo, and explain the takeover.
- [ ] If you intend to also make changes that are visible to the end user, like changing the command palette entry labels, you might want to consider creating a message for your new version to explain that to any existing users.

dennykorsukewitz commented:
Hi @braver,

Everything is ready, as far as I can see.

https://github.com/wbond/package_control_channel/pull/8904#issuecomment-2272292963

🚀

braver commented:
🎉

dennykorsukewitz commented:
Hi @braver 

do you have any idea why the package is not displayed on https://packagecontrol.io?

Best Denny

braver commented:
I think it’s this one?

https://packagecontrol.io/packages/Add%20Folder%20To%20Project

dennykorsukewitz commented:
Ahhh than I have to create a new PullRequest.

 Add Folder To Project is not assigned to existing user:
 https://packagecontrol.io/browse/authors/Denny%20Korsuk%C3%A9witz

My bad, i will fix that

Best Denny
